File: examples/javascript/fxbattleclient.js
'use strict';

import fetch from 'node-fetch';

export class FxBattleClient {
  constructor(endpoint, api_key){
    // the endpoint to connect to: e.g. 'http://localhost:8080'
    this.endpoint = endpoint;
    // the apikey to use: e.g. 'api_key2'
    this.api_key = api_key;
  }

  // helper function to handle the response from the api
  getJson(r) {
    if(r.status !== 200)
      throw "non 200 error code";
    if(r.headers.get('Content-Type') !== 'application/json')
      throw "unknown content type: " + r.headers.get('Content-Type');
    return r.json();
  }
  
  filterError(d) { 
    if(typeof(d) !== 'object') 
      throw "expected object";
    return d; 
  }
  
  account() {
    return fetch(this.endpoint + '/account/' + this.api_key)
      .then(this.getJson).then(this.filterError);
  }
  
  market() {
    return fetch(this.endpoint + '/market')
      .then(this.getJson).then(this.filterError);  
  }

  buy(ccy_pair, amount) {
    return fetch(this.endpoint + '/trade/' + this.api_key 
                 + '/buy/' + ccy_pair + '/' + amount )
      .then(this.getJson).then(this.filterError); 
  } 

  sell(ccy_pair, amount) {
    return fetch(this.endpoint + '/trade/' + this.api_key 
                 + '/sell/' + ccy_pair + '/' + amount )
      .then(this.getJson).then(this.filterError); 
  } 
}

File: examples/java/FXBattleClient.java
package fxbattle;

import com.mashape.unirest.http.*;
import com.mashape.unirest.http.exceptions.UnirestException;
import org.json.JSONObject;

public class FXBattleClient {
    private String endpoint;
    private String apiKey;

    public FXBattleClient(String endpoint, String apiKey) {
        this.endpoint = endpoint;
        this.apiKey = apiKey;
    }

    public JSONObject buy(String ccyPair, Double amount) throws UnirestException {
        return Unirest
            .get(this.endpoint + "/trade/{api_key}/buy/{ccy_pair}/{amount}")
            .routeParam("api_key", this.apiKey)
            .routeParam("ccy_pair", ccyPair)
            .routeParam("amount", amount.toString())
            .asJson().getBody().getObject();
    }

    public JSONObject sell(String ccyPair, Double amount) throws UnirestException {
        return Unirest
            .get(this.endpoint + "/trade/{api_key}/sell/{ccy_pair}/{amount}")
            .routeParam("api_key", this.apiKey)
            .routeParam("ccy_pair", ccyPair)
            .routeParam("amount", amount.toString())
            .asJson().getBody().getObject();
    }

    public JSONObject account() throws UnirestException {
        return Unirest
            .get(this.endpoint + "/account/{api_key}")
            .routeParam("api_key", this.apiKey)
            .asJson().getBody().getObject();
    }

    public JSONObject market() throws UnirestException {
        return Unirest
            .get(this.endpoint + "/market")
            .asJson().getBody().getObject();
    }
}

